### PR TITLE
Fixup nested step callbacks & non-hash input handling 

### LINKED ISSUE #350: Processor steps callbacks for nested schema are called in wrong order
## Describe the bug

```ruby
require 'dry-schema'
require 'json'

field_schema = Dry::Schema.define do
  required(:string_field)

  before(:key_coercer) { |result| result.to_h }
end

schema = Dry::Schema.define do
  required(:schema_field).filled(field_schema)

  before(:key_coercer) do |result|
    JSON.parse(result.output, {symbolize_names: true})
  end
end

p schema.call({ schema_field: { string_field: 'foo' } }.to_json)
```

## To Reproduce

Run example above, 
```
/home/ojab/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/dry-schema-1.6.2/lib/dry/schema/result.rb:52:in `[]': no implicit conversion of Symbol into Integer (TypeError)
	from /home/ojab/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/dry-schema-1.6.2/lib/dry/schema/result.rb:52:in `block in at'
	from /home/ojab/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/dry-schema-1.6.2/lib/dry/schema/path.rb:85:in `each'
	from /home/ojab/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/dry-schema-1.6.2/lib/dry/schema/path.rb:85:in `each'
	from /home/ojab/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/dry-schema-1.6.2/lib/dry/schema/result.rb:52:in `reduce'
…
```
We're not applying `schema.before(:key_coercer)` before trying to get `result[:schema_field]` for `field_schema`, which understandably fails since result contains unparsed string. 

## Expected behavior

It works! 

## My environment

- Affects my production application: NO
- Ruby version: 3.0.1
- OS: Linux x86_64

### PR DESCRIPTION
**Pull request by ojab**
Fixes #350 
Also some drive-by simplifications to `Path`

dry-v specs are passing with this changes.

### CODE REVIEW

File: lib/dry/schema/step.rb [hunk 1]
```diff
@@ -17,53 +17,30 @@ class Step
...
+          scoped_result.replace(output) if output.is_a?(Hash)
+          output
+        end
       end
 
       # @api private
-      def scoped(path)
-        Scoped.new(path, self)
+      def scoped(new_path)
+        path.keys.unshift(*new_path)
```
Comment (ojab) [inline lib/dry/schema/step.rb:42]: We're mutating step here, but right now it doesn't matter and easy change if would be needed later.

Comment (ojab) [reply lib/dry/schema/step.rb]: Ok, it actually breaks reused schemas, fixed.

**Comment by ojab** [comment]
Added handling of schemas inside arrays. Will break this to multiple PRs/squash/whatever if needed.

I don't quite like how it works, but dunno if it could be done undoubtly better.
IMHO would be easier and more clear to convert key_coercer/filter_schema/value_coercer to underlying type's constructor modifiers instead.

**Comment by ojab** [comment]
Found/fixed an issue with callbacks for missing `optional.schema`, also there is a noticeable hit in microbenchmarks after this PR, will take a look sometime this week.

**Comment by solnic** [comment]
Hey man thanks for this PR, it's awesome that you started tackling this. FYI `Path` and related things have been on a list of things to refactor/clean-up/improve for a very long time now. I'm more than happy to merge your stuff but please create separate PRs with your work split into refactors and bug fixes. This lib is foundational so we need to be more careful with changes.

**Comment by solnic** [comment]
@ojab so I guess this can be closed now, right? 🙂

**Comment by ojab** [comment]
Yeah, remaining part is array handling, I'll prepare a clean new PR with that. Thanks!

**Comment by solnic** [comment]
@ojab oh damn I thought everything was covered and I already released it 🙂

**Pull request closed without merging**